libavfilter/vf_signature.c

```diff
@@ -748,6 +758,310 @@ static int config_output(AVFilterLink *outlink)
+
+	// NumOfFrames
+	// it's the number of fine signatures
+	sc->lastindex = get_bits_long(&bitContext, 32);
+
+	// sc->time_base.den / sc->time_base.num
+	// hoping num is 1, other values are vague
+	// den/num might be greater than 16 bit, so cutting it
+	//put_bits(&buf, 16, 0xFFFF & (sc->time_base.den / sc->time_base.num));
+	// MediaTimeUnit
+
+	sc->time_base.den = get_bits(&bitContext, 16);
+	sc->time_base.num = 1;
+
+	// Skip the following data
+	// - MediaTimeFlagOfSpatialRegion: (1 bit) always 1
+	// - StartMediaTimeOfSpatialRegion: (32 bits) always 0
+	skip_bits(&bitContext, 1 + 32);
+
+	// EndMediaTimeOfSpatialRegion
+	uint64_t lastCoarsePts = get_bits_long(&bitContext, 32);
+
+	// Coarse signatures
+	// numOfSegments = number of coarse signatures
+	numOfSegments = get_bits_long(&bitContext, 32);
+
+	// Reading from binary signature return a wrong number of segments
+	// numOfSegments = (sc->lastindex + 44)/45;
+	//skip_bits(&bitContext, 32);
+
+	sc->coarsesiglist = (CoarseSignature*)calloc(numOfSegments,	sizeof(CoarseSignature));
+
+	BoundedCoarseSignature *bCoarseList = (BoundedCoarseSignature*)calloc(numOfSegments, sizeof(BoundedCoarseSignature));
+
+	// CoarseSignature loading
+	for (unsigned int i = 0; i < numOfSegments; ++i) {
+		BoundedCoarseSignature *bCs = &bCoarseList[i];
+		bCs->cSign = &sc->coarsesiglist[i];
+
+		if (i < numOfSegments - 1)
+			bCs->cSign->next = &sc->coarsesiglist[i + 1];
+		// each coarse signature is a VSVideoSegment
+		// StartFrameOfSegment
+		bCs->firstIndex = get_bits_long(&bitContext, 32);
+		// EndFrameOfSegment
+		bCs->lastIndex = get_bits_long(&bitContext, 32);
+
+		// MediaTimeFlagOfSegment 1 bit, always 1
+		skip_bits(&bitContext, 1);
+
+		// Fine signature pts
+		// StartMediaTimeOfSegment 32 bits
+		bCs->firstPts = get_bits_long(&bitContext, 32);
+		// EndMediaTimeOfSegment 32 bits
+		bCs->lastPts = get_bits_long(&bitContext, 32);
+		// Bag of words
+		for (unsigned int i = 0; i < 5; ++i) {
+
+			// read 243 bits ( = 7 * 32 + 19 = 8 * 28 + 19) into buffer
+			for (unsigned int j = 0; j < 30; ++j) {
+				// 30*8 bits = 30 bytes
+				bCs->cSign->data[i][j] = get_bits(&bitContext, 8);
+			}
+			bCs->cSign->data[i][30] = get_bits(&bitContext, 3) << 5;
+		}
+	}
+	sc->coarseend = &sc->coarsesiglist[numOfSegments - 1];
+
+	// Finesignatures
+	// CompressionFlag, only 0 supported
+	skip_bits(&bitContext, 1);
+
+	sc->finesiglist = (FineSignature*)calloc(sc->lastindex,	sizeof(FineSignature));
+
+	// Load fine signatures from file
+	for (unsigned int i = 0; i < sc->lastindex; ++i) {
+		FineSignature *fs = &sc->finesiglist[i];
+
+		// MediaTimeFlagOfFrame always 1
+		skip_bits(&bitContext, 1);
+
+		// MediaTimeOfFrame (PTS)
+		fs->pts = get_bits_long(&bitContext, 32);
+
+		// FrameConfidence
+		fs->confidence = get_bits(&bitContext, 8);
+
+		// words
+		for (unsigned int l = 0; l < 5; l++) {
+			fs->words[l] = get_bits(&bitContext, 8);
+		}
+
+		// Crashes for some signature, it's a memory adding problems
```

Comment:
Hmm I guess not every frame out of `sc->lastindex` has a valid frame signature in the output binary. We should think about alternatives here.

I see that the `get_bits.h` utility in ffmpeg offers a safe bound-checking API by setting a global macro https://github.com/livepeer/FFmpeg/blob/b344789a31a2d806be702239d7614f7a9f51f941/libavcodec/get_bits.h#L41 - but I think doing that for our whole ffmpeg compile might slow down other demuxers.

Maybe we can calculate the total file size we've read beforehand, and do bound checking ourselves comparing it with `GetBitContext->index`